File: cricrelay-mobile/ios/Features/Studio/StudioChecklist.swift
import Foundation

// MARK: - 1b Checklist gate — the go-live gate IS the setup UI.
//
// Three checks (camera / destination / scoreboard source) derived from published
// StudioViewModel state. The segmented Go Live ring shows N/3, the checklist rows are
// the sheet entry points, and a blocked ring routes the operator to the first
// incomplete check instead of going dead. Mirrors Android's StudioChecklist.kt.

/// Which of the three readiness checks a row represents; tapping opens its sheet.
enum CheckKind: Hashable {
    case camera
    case destination
    case scoring
}

struct StudioCheck: Identifiable, Equatable {
    let kind: CheckKind
    let complete: Bool
    let title: String
    let sublabel: String
    /// Paint the sublabel coral (guidance/stale states) instead of the dim body colour.
    let warning: Bool

    var id: CheckKind { kind }
}

enum StudioChecklist {
    /// Pure derivation so it stays unit-testable — every input is plain state.
    static func build(
        previewReady: Bool,
        qualityLabel: String,
        stabilizationLevel: Int,
        destinationReady: Bool,
        destinationLabel: String,
        scoringMode: String,
        overlayEmbedUrl: String,
        matchDay: MatchDayStatus?
    ) -> [StudioCheck] {
        [
            cameraCheck(
                previewReady: previewReady,
                qualityLabel: qualityLabel,
                stabilizationLevel: stabilizationLevel
            ),
            destinationCheck(ready: destinationReady, label: destinationLabel),
            scoringCheck(mode: scoringMode, overlayEmbedUrl: overlayEmbedUrl, matchDay: matchDay),
        ]
    }

    static func completedCount(_ checks: [StudioCheck]) -> Int {
        checks.filter(\.complete).count
    }

    static func firstIncomplete(_ checks: [StudioCheck]) -> StudioCheck? {
        checks.first { !$0.complete }
    }

    /// One line under the ring naming the missing check ("blocked = guidance, never dead").
    static func ringCaption(_ checks: [StudioCheck]) -> String {
        guard let first = firstIncomplete(checks) else {
            return "All checks passed — you're good to go"
        }
        switch first.kind {
        case .camera: return "Waiting for the camera to be ready"
        case .destination: return "Choose a destination to unlock"
        case .scoring: return "Choose a scoring source to unlock"
        }
    }

    /// Coral "N to fix" inside the blocked ring; nil once every check passes.
    static func fixLabel(_ checks: [StudioCheck]) -> String? {
        let missing = checks.count - completedCount(checks)
        return missing == 0 ? nil : "\(missing) to fix"
    }

    // MARK: - Individual checks

    private static func cameraCheck(
        previewReady: Bool,
        qualityLabel: String,
        stabilizationLevel: Int
    ) -> StudioCheck {
        let stabName: String
        switch stabilizationLevel {
        case StabilizationLevel.cinematic.rawValue: stabName = "cinematic"
        case StabilizationLevel.off.rawValue: stabName = "off"
        default: stabName = "standard"
        }
        return StudioCheck(
            kind: .camera,
            complete: previewReady,
            title: previewReady ? "Camera ready" : "Camera",
            sublabel: previewReady
                ? "\(qualityLabel) · stabilization \(stabName)"
                : "Waiting for the camera…",
            warning: false
        )
    }

    private static func destinationCheck(ready: Bool, label: String) -> StudioCheck {
        StudioCheck(
            kind: .destination,
            complete: ready,
            title: ready ? "\(label) connected" : "Destination",
            sublabel: ready ? "Ready to stream" : "Choose where to stream",
            warning: !ready
        )
    }

    private static func scoringCheck(
        mode: String,
        overlayEmbedUrl: String,
        matchDay: MatchDayStatus?
    ) -> StudioCheck {
        let modeSet = !mode.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        let overlayLinked = !overlayEmbedUrl.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        // Both halves preserve the old preflight's coverage: a scoring mode without an overlay
        // embed URL means going live silently drops the board from the stream.
        let complete = modeSet && overlayLinked

        if !complete {
            return StudioCheck(
                kind: .scoring,
                complete: false,
                title: "Scoreboard source",
                sublabel: modeSet
                    ? "Overlay link missing — re-link the match"
                    : "Nothing linked yet",
                warning: true
            )
        }

        // Sublabel from the live MatchDayStatus poll (~5 s cadence).
        let effectiveMode = matchDay?.scoringMode ?? mode
        if effectiveMode == "manual" {
            let sublabel: String
            var warning = false
            if matchDay?.scoringStale == true {
                sublabel = "Manual scorer · updates stalled"; warning = true
            } else if matchDay?.scoringActive == true {
                sublabel = "Manual scorer · updating"
            } else {
                sublabel = "Manual scorer · waiting for scorer"
            }
            return StudioCheck(
                kind: .scoring, complete: true,
                title: "Scoreboard linked", sublabel: sublabel, warning: warning
            )
        }
        if let day = matchDay, day.scoringStale {
            return StudioCheck(
                kind: .scoring, complete: true,
                title: "Scoreboard linked", sublabel: "Play-Cricket · feed stale", warning: true
            )
        }
        let live = matchDay?.scoringActive == true
        return StudioCheck(
            kind: .scoring, complete: true,
            title: "Scoreboard linked",
            sublabel: live ? "Play-Cricket · live" : "Play-Cricket · waiting for feed",
            warning: false
        )
    }
}

// MARK: - View-model derivation

extension StudioViewModel {
    /// The three readiness checks the 1b gate renders, derived from published state so
    /// SwiftUI recomputes them on any relevant change.
    var checks: [StudioCheck] {
        StudioChecklist.build(
            previewReady: previewReady,
            qualityLabel: Self.streamQualityLabel,
            stabilizationLevel: overlayPrefs.stabilizationLevel,
            destinationReady: destinationReady,
            destinationLabel: destinationLabel,
            scoringMode: scoringConfig?.mode ?? "",
            overlayEmbedUrl: match?.overlayEmbedUrl ?? "",
            matchDay: matchDay
        )
    }

    var completedChecksCount: Int { StudioChecklist.completedCount(checks) }

    var firstIncompleteCheck: StudioCheck? { StudioChecklist.firstIncomplete(checks) }

    var ringCaption: String { StudioChecklist.ringCaption(checks) }

    var goLiveFixLabel: String? { StudioChecklist.fixLabel(checks) }

    /// Configured stream quality ("1080p30"). The iOS engine publishes no measured bitrate,
    /// so the health pill shows this configured quality only.
    /// TODO: surface measured bitrate when the engine publishes stats.
    static var streamQualityLabel: String {
        "\(min(StreamCameraEngine.defaultStreamWidth, StreamCameraEngine.defaultStreamHeight))p30"
    }

    /// Route a checklist row (or a blocked ring tap) to the sheet that fixes it.
    func openSheet(for kind: CheckKind) {
        switch kind {
        case .camera: activeSheet = .cameraSettings
        case .destination: activeSheet = .destination
        case .scoring: activeSheet = .scoring
        }
    }
}

File: cricrelay-mobile/android/feature/studio/src/main/java/uk/co/cricrelay/mobile/feature/studio/StudioChecklist.kt
package uk.co.cricrelay.mobile.feature.studio

import uk.co.cricrelay.shared.model.StabilizationLevel

/** Which of the three go-live checks a checklist row represents (1b Checklist gate). */
enum class CheckKind { Camera, Destination, Scoring }

/**
 * One row of the go-live checklist — pure data derived from [StudioUiState] so the gate is
 * unit-testable without Compose. [warning] rows render the coral blocked-as-guidance
 * treatment (tinted row + "Choose" chip); complete rows get the sky check disc.
 */
data class StudioCheck(
    val kind: CheckKind,
    val complete: Boolean,
    val title: String,
    val sublabel: String,
    val warning: Boolean = !complete,
)

/**
 * Derivation for the 1b Checklist gate: the segmented Go Live ring and the checklist panel
 * both read from here, so "what blocks Go Live" has exactly one definition. Check 3 folds in
 * the old preflight's overlay-URL coverage — a configured scorer with no overlay embed would
 * otherwise go live with a silent, board-less stream.
 */
object StudioChecklist {

    fun deriveChecks(state: StudioUiState): List<StudioCheck> = listOf(
        cameraCheck(state),
        destinationCheck(state),
        scoringCheck(state),
    )

    fun completedCount(checks: List<StudioCheck>): Int = checks.count { it.complete }

    fun firstIncomplete(checks: List<StudioCheck>): StudioCheck? =
        checks.firstOrNull { !it.complete }

    /** Caption under the ring — names the first missing check per SPEC blocked-as-guidance. */
    fun ringCaption(checks: List<StudioCheck>): String = when (firstIncomplete(checks)?.kind) {
        null -> "All checks passed — tap to go live"
        CheckKind.Camera -> "Waiting for the camera to unlock"
        CheckKind.Destination -> "Choose a destination to unlock"
        CheckKind.Scoring -> "Choose a scoring source to unlock"
    }

    /** Coral "N to fix" line inside the blocked ring; null once every check passes. */
    fun fixLabel(checks: List<StudioCheck>): String? {
        val remaining = checks.count { !it.complete }
        return if (remaining == 0) null else "$remaining to fix"
    }

    private fun cameraCheck(state: StudioUiState): StudioCheck {
        if (!state.previewReady) {
            return StudioCheck(
                kind = CheckKind.Camera,
                complete = false,
                title = "Camera",
                sublabel = "Waiting for the preview…",
            )
        }
        val quality = if (state.supports1080p) "1080p30" else "720p30"
        val stabilization = when (state.overlayPrefs.stabilizationLevel) {
            StabilizationLevel.CINEMATIC -> "cinematic"
            StabilizationLevel.STANDARD -> "standard"
            else -> "off"
        }
        return StudioCheck(
            kind = CheckKind.Camera,
            complete = true,
            title = "Camera ready",
            sublabel = "$quality · stabilization $stabilization",
        )
    }

    private fun destinationCheck(state: StudioUiState): StudioCheck {
        if (!state.destinationReady) {
            return StudioCheck(
                kind = CheckKind.Destination,
                complete = false,
                title = "Destination",
                sublabel = "Nowhere to stream yet",
            )
        }
        val (title, sublabel) = when (state.destination) {
            StreamDestination.YouTube -> "YouTube connected" to "Club channel via OAuth"
            StreamDestination.Twitch -> "Twitch connected" to "Club channel via OAuth"
            StreamDestination.Custom -> "Custom RTMP set" to (
                if (state.selectedSavedDestinationId != null) "Saved destination ready"
                else "Server URL and stream key saved"
            )
        }
        return StudioCheck(CheckKind.Destination, complete = true, title = title, sublabel = sublabel)
    }

    private fun scoringCheck(state: StudioUiState): StudioCheck {
        val configured = state.scoring?.mode?.isNotBlank() == true
        val overlayOk = state.match?.overlayEmbedUrl?.isNotBlank() == true
        return when {
            !configured -> StudioCheck(
                kind = CheckKind.Scoring,
                complete = false,
                title = "Scoreboard source",
                sublabel = "Nothing linked yet",
            )
            // Preserves the old preflight's coverage: a linked scorer is useless if the
            // stream has no overlay embed URL to rasterize.
            !overlayOk -> StudioCheck(
                kind = CheckKind.Scoring,
                complete = false,
                title = "Scoreboard source",
                sublabel = "Overlay URL missing — check the stream setup",
            )
            else -> StudioCheck(
                kind = CheckKind.Scoring,
                complete = true,
                title = "Scoreboard source",
                sublabel = scoringSublabel(state),
            )
        }
    }

    private fun scoringSublabel(state: StudioUiState): String {
        val mode = state.matchDay?.scoringMode ?: state.scoring?.mode.orEmpty()
        val base = when {
            mode.equals("auto", ignoreCase = true) -> "Auto (Play-Cricket)"
            mode.equals("manual", ignoreCase = true) -> "Manual scorer"
            mode.isBlank() -> "Linked"
            else -> mode.replaceFirstChar { it.uppercase() }
        }
        return when {
            state.matchDay?.scoringStale == true -> "$base · updates stalled"
            state.matchDay?.scoringActive == true -> "$base · updating"
            mode.equals("manual", ignoreCase = true) &&
                state.matchDay != null -> "$base · waiting for scorer"
            else -> base
        }
    }
}
